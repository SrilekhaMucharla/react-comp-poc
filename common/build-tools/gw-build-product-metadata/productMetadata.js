const _ = require('lodash');
const { readJson } = require('fs-extra');
const appPaths = require('gw-build-config-paths');
const { sendJsonRpcRequest } = require('gw-build-jsonrpc');
const fileUtils = require('./util/fileUtils');
const metadataLocalisation = require('./metadataLocalisation');

function toMetadataMap(serverMetadataResult) {
    const metadataByName = _.groupBy(serverMetadataResult, 'name');
    return _.mapValues(metadataByName, (typeMetas) => {
        const allMetadataForName = _.map(typeMetas, 'metadata');
        return _.merge({}, ...allMetadataForName);
    });
}

/**
 * Fetches the metadata for the given endpoint
 *
 * @param {ApplicationEndpoint} endpoint the endpoint which metadata should be fetched
 * @returns {Promise<Object>} the metadata retrieved from the given endpoint
 */
function fetchMetadataForEndpoint(endpoint) {
    return sendJsonRpcRequest(endpoint, 'getMetaData')
        .then((metadata) => ({
            endpoint,
            metadata
        }));
}

/**
 * @typedef { import("gw-build-capabilities-management").ApplicationEndpoint } ApplicationEndpoint
 */

/**
 * @typedef PreprocessMetadata
 * @property {{server: string, service: string}} endpoint the server nickname (e.g. pc, cc, ...)
 * @property {object} metadata the metadata from the server
 */

/**
 * Structures the product metadata
 *
 * @param {PreprocessMetadata} data containing the endpoint from which
 * the metadata came from and the metadata from the server
 * @returns {Object} the metadata retrieved from the given endpoint
 */
function preprocessDtoMetadata({ endpoint, metadata }) {
    return {
        [endpoint.server]: toMetadataMap(metadata.newTypeMeta)
    };
}

/**
 * Structures the question set metadata
 *
 * @param {PreprocessMetadata} data containing the endpoint from which
 * the metadata came from and the metadata from the server
 * @returns {Object} the metadata from the server
 */
function preprocessQuestionSetsMetadata({ metadata }) {
    return metadata.questionSets;
}

/**
 * Combines all the metadatas from all the endpoints into one object
 *
 * @param {Array<Object>} allMetadataResponses containing the metadata from the server
 * @returns {Object} the metadata to save
 */
function mergeMetadata(allMetadataResponses) {
    return allMetadataResponses.reduce((generatedMetadata, serviceMetadata) => {
        return _.merge(generatedMetadata, serviceMetadata);
    }, {});
}

/**
 * Pipeline to format, combine and save the metadata
 *
 * @param {Promise<Object>} metadataPromises response from the server
 * @param {function} preprocessingFunction function to format the metadata from the server
 * @param {string} destinationFile path to save the metadata
 * @returns {Object} the metadata to save
 */
function metadataPipeline(metadataPromises, preprocessingFunction, destinationFile) {
    return metadataPromises
        .then((metadataFromEndpoints) => metadataFromEndpoints.map(preprocessingFunction))
        .then(mergeMetadata)
        .then((allMetadata) => fileUtils.writeFile(allMetadata, destinationFile));
}

/**
 * Process the product metadata
 *
 * @param {Promise<Object>} metadataPromises response from the server
 * @returns {Promise} the metadata to save
 */
function processDTOMetadata(metadataPromises) {
    return metadataPipeline(metadataPromises, preprocessDtoMetadata, appPaths.appProductMetadata);
}

/**
 * Process the question sets metadata
 *
 * @param {Promise<Object>} metadataPromises response from the server
 * @returns {Promise} the metadata to save
 */
function processQuestionSetsMetadata(metadataPromises) {
    return metadataPipeline(metadataPromises, preprocessQuestionSetsMetadata,
        appPaths.appQuestionSetsMetadata);
}

async function buildMetadata() {
    const { capabilities } = await readJson(appPaths.appGeneratedConfig);
    const endpoints = Object.values(capabilities);
    const endpointsWithMetadata = endpoints.filter((endpoint) => !endpoint.skipMetadata);
    const metadataPromises = Promise.all(endpointsWithMetadata.map(fetchMetadataForEndpoint));
    await Promise.all([
        processDTOMetadata(metadataPromises),
        processQuestionSetsMetadata(metadataPromises)
    ]);
    const localeConfig = await metadataLocalisation.fetchLocaleInfo();
    metadataLocalisation.processMetadataDisplayKeys(localeConfig, endpoints);
}

module.exports = {
    buildMetadata
};
