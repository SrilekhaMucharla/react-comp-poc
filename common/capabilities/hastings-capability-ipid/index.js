// eslint-disable-next-line import/prefer-default-export
export { default as HastingsIpidService } from './services/HastingsIpidService';
