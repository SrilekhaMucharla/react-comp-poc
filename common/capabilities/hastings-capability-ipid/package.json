{
  "name": "hastings-capability-ipid",
  "description": "Provides capability to manage ipid doc",
  "main": "index.js",
  "version": "1.0.0",
  "author": "Hastings",
  "private": true,
  "dependencies": {},
  "peerDependencies": {
    "gw-portals-transport-js": "11.2.0-db467149eb42",
    "gw-portals-url-js": "11.2.0-db467149eb42"
  }
}
