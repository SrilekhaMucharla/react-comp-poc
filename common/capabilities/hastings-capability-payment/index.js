// eslint-disable-next-line import/prefer-default-export
export { default as HastingsPaymentService } from './services/HastingsPaymentService';
