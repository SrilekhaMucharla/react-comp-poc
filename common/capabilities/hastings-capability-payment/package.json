{
  "name": "hastings-capability-payment",
  "description": "Provides capability to payment operations",
  "main": "index.js",
  "version": "1.0.0",
  "author": "Hastings",
  "private": true,
  "dependencies": {},
  "peerDependencies": {
    "gw-portals-transport-js": "11.2.0-db467149eb42",
    "gw-portals-url-js": "11.2.0-db467149eb42"
  }
}
