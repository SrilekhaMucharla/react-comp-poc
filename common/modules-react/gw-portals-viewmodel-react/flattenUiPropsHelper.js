/* eslint-disable */
import _ from "lodash";

/**
 *
 * @param {object} uiProps
 */
export function getFlattenedUiPropsContent(uiProps) {
  if (uiProps.contentRepeat) {
    return uiProps;
  }

  if (!_.isArray(uiProps.content)) {
    return uiProps;
  }

  return uiProps.content.flatMap((item) => getFlattenedUiPropsContent(item));
}
