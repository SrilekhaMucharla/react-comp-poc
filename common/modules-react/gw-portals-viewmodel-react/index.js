export { default as ViewModelServiceContext } from './ViewModelServiceContext';
export { default as withViewModelService } from './withViewModelServiceHOC';
export { getFlattenedUiPropsContent } from './flattenUiPropsHelper';
