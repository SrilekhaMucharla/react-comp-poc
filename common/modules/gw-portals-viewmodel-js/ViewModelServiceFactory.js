import InsuranceSuiteMetadataServiceFactory from 'gw-portals-metadata-js';
import DateUtil from 'gw-portals-util-js/DateUtil';
import LocalDateUtil from 'gw-portals-util-js/LocalDateUtil';

import ViewModel from './ViewModel';

import Context from './aspects/Context';
import Requiredness from './aspects/Requiredness';
import Validity from './aspects/Validity';
import SubtreeValidity from './aspects/SubtreeValidity';
import InputCtrlType from './aspects/InputCtrlType';
import AvailableValues from './aspects/AvailableValues';
import SerializedForm from './aspects/SerializedForm';
import RebuildVMFromData from './aspects/RebuildVMFromData';
import DebugInvalidChildProperties from './aspects/DebugInvalidChildProperties';
import TypedValue from './aspects/TypedValue';
import _ValidationFunctions from './aspects/_ValidationFunctions';
import _AddressAutocompleteUtil from './aspects/_AddressAutocompleteUtil';

import { ExpressionLanguageServiceFactory } from './expression-language';
import _PhoneUtil from './util/PhoneUtil';
import DateFormatter from './util/DateFormatter';

function getViewModelService(metadata, translator) {
    const suiteMetadata = InsuranceSuiteMetadataServiceFactory.createSuiteMetadata(metadata);
    const ExpressionLanguageService = ExpressionLanguageServiceFactory.get(
        { insuranceSuiteMetadataService: suiteMetadata, translator }
    );
    // view model initiation which is done in the initial module in Portals (extracted)
    const ViewModelService = ViewModel.get(suiteMetadata);

    const typeValidationMessages = {};
    typeValidationMessages.notANumber = translator('platform.modelvalidation.DTOValidationService.Please enter numbers only');
    typeValidationMessages.notAnInteger = translator('platform.modelvalidation.DTOValidationService.Value entered must not contain decimal values');

    ViewModelService.addAspectFactory(Context.create(ExpressionLanguageService));
    ViewModelService.addAspectFactory(Requiredness.create(ExpressionLanguageService));
    ViewModelService.addAspectFactory(
        Validity.create(ExpressionLanguageService, typeValidationMessages)
    );
    ViewModelService.addAspectFactory(SubtreeValidity);
    ViewModelService.addAspectFactory(InputCtrlType);
    ViewModelService.addAspectFactory(AvailableValues.create(ExpressionLanguageService));
    ViewModelService.addAspectFactory(SerializedForm);
    ViewModelService.addAspectFactory(RebuildVMFromData);
    ViewModelService.addAspectFactory(DebugInvalidChildProperties);
    ViewModelService.addAspectFactory(TypedValue);

    ExpressionLanguageService.getCompilationContext().registerType('edgev10.aspects.validation.ValidationFunctions', _ValidationFunctions());
    ExpressionLanguageService.getCompilationContext().registerType('gw.api.contact.AddressAutocompleteUtil', _AddressAutocompleteUtil());
    ExpressionLanguageService.getCompilationContext().registerType('edgev10.capabilities.locale.util.DateFormatter', DateFormatter);
    ExpressionLanguageService.getCompilationContext().registerType('gw.api.util.DateUtil', DateUtil);
    ExpressionLanguageService.getCompilationContext().registerType('gw.api.util.PhoneUtil', _PhoneUtil());
    ExpressionLanguageService.getCompilationContext().registerType('edgev10.time.LocalDateUtil', LocalDateUtil);

    return {
        create: ViewModelService.create,
        changeContext: ViewModelService.changeContext,
        clone: ViewModelService.clone
    };
}

// EXPORT
export default {
    getViewModelService
};
