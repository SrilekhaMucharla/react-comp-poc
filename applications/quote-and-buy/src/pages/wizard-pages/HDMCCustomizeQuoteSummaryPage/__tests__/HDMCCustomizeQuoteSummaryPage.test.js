import React from 'react';
import { shallow, mount } from 'enzyme';
import { Router } from 'react-router-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import _ from 'lodash';
import ViewModelServiceFactory from 'gw-portals-viewmodel-js/ViewModelServiceFactory';
import { ViewModelServiceContext } from 'gw-portals-viewmodel-react';
import { HastingsIpidService } from 'hastings-capability-ipid';
import { HastingsPaymentService } from 'hastings-capability-payment';
import rootReducer from '../../../../redux-thunk/reducers/index';
import defaultTranslator from '../../__helpers__/testHelper';
import productMetadata from '../../../../generated/metadata/product-metadata.json';
import mcsubmission from '../mock/retrieveResponse.json';
import submission from '../../HDCustomizeQuoteSummaryPage/mock/mockSubmission.json';
import HDMCCustomizeQuoteSummaryPageConnected from '../HDMCCustomizeQuoteSummaryPage';
import withTranslator from '../../__helpers__/test/withTranslator';
import mockPaymentSchedule from '../../HDMCThanksPage/mock/mockMCPSnormal.json';
// import routes from '../../../../routes/WizardRouter/RouteConst';

jest.mock('react-dom', () => ({
    createPortal: (node) => node,
}));

jest.mock('../../../../../../../node_modules/react-redux/lib/utils/batch.js', () => ({
    setBatch: jest.fn(),
    getBatch: () => (fn) => fn()
}));

jest.spyOn(HastingsIpidService, 'mcIpidByProducerCode').mockResolvedValue({ result: { ipidResponse: [{ ipids: {} }] } });

// const VEHICLE_PATH = routes.VEHICLE_DETAILS;
// const DRIVER_PATH = routes.MC_MILESTONE;

const viewModelService = ViewModelServiceFactory.getViewModelService(
    productMetadata, defaultTranslator
);

function createInitialState() {
    const MCsubmissionVM = viewModelService.create(
        mcsubmission,
        'pc',
        'com.hastings.edgev10.capabilities.quote.submission.dto.HastingsMultiQuoteDataDTO'
    );

    const SubmissionVM = viewModelService.create(
        submission,
        'pc',
        'edgev10.capabilities.quote.submission.dto.QuoteDataDTO'
    );

    const initialState = {
        wizardState: {
            data: {
                mcsubmissionVM: MCsubmissionVM,
                submissionVM: SubmissionVM
            },
            app: {
                step: 1,
                prevStep: 0
            },
        },
        rerateModal: { status: false },
        customMultiQuoteModel: {
            loading: true,
            multiCustomUpdatedQuoteObj: {}
        },
        mcPaymentScheduleModel: {
            mcPaymentScheduleObject: mockPaymentSchedule.result,
            mcPaymentScheduleError: {}
        }
    };

    return initialState;
}

// actual store since component relies on dispatch to e.g., initialize createSumbissionVM to store
function initializeRealStore(...initialStateModifiers) {
    const initialState = createInitialState();

    // apply initialState modifiers
    _.over(initialStateModifiers)(initialState);

    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk)
    );
    return store;
}

async function initializeWrapper(store, props, history = createMemoryHistory()) {
    let wrapper;
    await act(async () => {
        wrapper = mount(withTranslator(
            <ViewModelServiceContext.Provider value={viewModelService}>
                <Router history={history}>
                    <Provider store={store}>
                        <HDMCCustomizeQuoteSummaryPageConnected
                            setTotalPrice={jest.fn()}
                            shouldPNCDShow={jest.fn()}
                            {...props} />
                    </Provider>
                </Router>
            </ViewModelServiceContext.Provider>
        ));
    });
    wrapper.update();
    return wrapper;
}

describe('<HDMCCustomizeQuoteSummaryPage />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.spyOn(HastingsPaymentService, 'fetchMCPaymentDetails').mockImplementation((data) => Promise.resolve({
            result: data
        }));
    });

    test('render component', () => {
        const wrapper = shallow((
            <HDMCCustomizeQuoteSummaryPageConnected />
        ));

        expect(wrapper).toMatchSnapshot();
    });

    test('component renders', async () => {
        const store = initializeRealStore();
        const wrapper = await initializeWrapper(store);
        HastingsPaymentService.fetchMCPaymentDetails.mockResolvedValue(mockPaymentSchedule);
        expect(wrapper).toEqual({});
    });
});
