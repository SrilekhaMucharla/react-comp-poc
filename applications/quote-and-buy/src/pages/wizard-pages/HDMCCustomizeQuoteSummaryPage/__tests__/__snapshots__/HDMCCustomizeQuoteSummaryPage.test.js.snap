// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HDMCCustomizeQuoteSummaryPage /> render component 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
